Release pipelines failing with SaaS environments due BcAuthContext

Since this morning, our release pipelines have been failing when deploying to SaaS environments. The issue appears to be related to the Publish-PerTenantExtensionApps function, which was recently modified.
We are using BCContainerHelper version 6.0.26.


2024-10-16T11:46:24.2117258Z ##[section]Starting: Publish to PRE
2024-10-16T11:46:24.2126338Z ==============================================================================
2024-10-16T11:46:24.2126495Z Task         : PowerShell
2024-10-16T11:46:24.2126590Z Description  : Run a PowerShell script on Linux, macOS, or Windows
2024-10-16T11:46:24.2126704Z Version      : 2.245.1
2024-10-16T11:46:24.2126784Z Author       : Microsoft Corporation
2024-10-16T11:46:24.2127578Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
2024-10-16T11:46:24.2127710Z ==============================================================================
2024-10-16T11:46:26.1655676Z Generating script.
2024-10-16T11:46:26.3058028Z ========================== Starting Command Output ===========================
2024-10-16T11:46:26.3430541Z ##[command]"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'D:\a\_temp\c42c46eb-3db5-45ec-8319-43215f084752.ps1'"
2024-10-16T11:46:26.7391821Z Installing BcContainerHelper
2024-10-16T11:48:26.2312440Z Publishing to tenant
2024-10-16T11:48:26.7360427Z BcContainerHelper version 6.0.26
2024-10-16T11:48:29.3622866Z BC.HelperFunctions emits usage statistics telemetry to Microsoft
2024-10-16T11:48:29.5237858Z Running on Windows, PowerShell 5.1.20348.2652
2024-10-16T11:48:32.6017777Z LatestGenericTagVersion is 1.0.2.50
2024-10-16T11:48:33.8023667Z Publish-PerTenantExtensionApps Telemetry Correlation Id: 63187a89-017c-4e0e-9758-6995a0db55c4
2024-10-16T11:48:33.9056219Z BcAuthContext should be a HashTable created by New-BcAuthContext.
2024-10-16T11:48:33.9056814Z At C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\6.0.26\HelperFunctions.ps1:882 char:9
2024-10-16T11:48:33.9058505Z +         throw 'BcAuthContext should be a HashTable created by New-BcA ...
2024-10-16T11:48:33.9059542Z +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2024-10-16T11:48:33.9060487Z     + CategoryInfo          : OperationStopped: (BcAuthContext s...-BcAuthContext.:String) [], RuntimeException
2024-10-16T11:48:33.9062206Z     + FullyQualifiedErrorId : BcAuthContext should be a HashTable created by New-BcAuthContext.
2024-10-16T11:48:33.9062796Z  
2024-10-16T11:48:34.0678695Z ##[error]PowerShell exited with code '1'.
2024-10-16T11:48:34.1192899Z ##[section]Finishing: Publish to PRE

--- a/AppHandling/Compile-AppInNavContainer.ps1
+++ b/AppHandling/Compile-AppInNavContainer.ps1
@@ -435,7 +435,7 @@ try {
         $addDependencies = @()
         if ($existingApp) {
             Write-Host "Dependency App exists"
-            if ($existingApp.PropagateDependencies) {
+            if ($existingApp.ContainsKey('PropagateDependencies') -and $existingApp.PropagateDependencies -and $existingApp.ContainsKey('Dependencies')) {
                 $addDependencies += $existingApp.Dependencies
             }
         }

--- a/AppHandling/Compile-AppInNavContainer.ps1
+++ b/AppHandling/Compile-AppInNavContainer.ps1
@@ -431,7 +431,7 @@ try {
             else {
                 (($_.Name -eq $dependency.name) -and ($_.Name -eq "Application" -or (($_.Publisher -eq $dependency.publisher) -and ([System.Version]$_.Version -ge [System.Version]$dependency.version))))
             }
-        }
+        } | Sort-Object { [System.Version]$_.Version } -Descending | Select-Object -First 1
         $addDependencies = @()
         if ($existingApp) {
             Write-Host "Dependency App exists"

--- a/CompilerFolderHandling/Compile-AppWithBcCompilerFolder.ps1
+++ b/CompilerFolderHandling/Compile-AppWithBcCompilerFolder.ps1
@@ -186,7 +186,7 @@ try {
         Write-Host "Processing dependency $($dependency.Publisher)_$($dependency.Name)_$($dependency.Version) ($($dependency.AppId))"
         $existingApp = $existingApps | Where-Object {
             ((($dependency.appId -ne '' -and $_.AppId -eq $dependency.appId) -or ($dependency.appId -eq '' -and $_.Name -eq $dependency.Name)) -and ([System.Version]$_.Version -ge [System.Version]$dependency.version))
-        }
+        } | Sort-Object { [System.Version]$_.Version } -Descending | Select-Object -First 1
         $addDependencies = $()
         if ($existingApp) {
             Write-Host "Dependency App exists"

--- a/CompilerFolderHandling/Compile-AppWithBcCompilerFolder.ps1
+++ b/CompilerFolderHandling/Compile-AppWithBcCompilerFolder.ps1
@@ -190,7 +190,7 @@ try {
         $addDependencies = $()
         if ($existingApp) {
             Write-Host "Dependency App exists"
-            if ($existingApp.ContainsKey('PropagateDependencies') -and $existingApp.PropagateDependencies) {
+            if ($existingApp.ContainsKey('PropagateDependencies') -and $existingApp.PropagateDependencies -and $existingApp.ContainsKey('Dependencies')) {
                 $addDependencies += $existingApp.Dependencies
             }
         }

--- a/ReleaseNotes.txt
+++ b/ReleaseNotes.txt
@@ -1,5 +1,6 @@
 6.0.27
 Issue 3538 Compile-AppWithBcCompilerFolder fails when dependency does propagateDependencies
+Issue 3727 Regression - Release pipelines failing with SaaS environments due BcAuthContext
 
 6.0.26
 As minimum, always use the generic tag version which was available when shipping BcContainerHelper

--- a/Saas/Publish-PerTenantExtensionApps.ps1
+++ b/Saas/Publish-PerTenantExtensionApps.ps1
@@ -51,8 +51,6 @@ function Publish-PerTenantExtensionApps {
 $telemetryScope = InitTelemetryScope -name $MyInvocation.InvocationName -parameterValues $PSBoundParameters -includeParameters @()
 try {
 	
-    $script:authContext = Renew-BcAuthContext -bcAuthContext $bcAuthContext
-
     function GetAuthHeaders {
         $script:authContext = Renew-BcAuthContext -bcAuthContext $script:authContext
         return @{ "Authorization" = "Bearer $($script:authContext.AccessToken)" }
@@ -69,15 +67,18 @@ try {
         if ($clientSecret -is [String]) { $clientSecret = ConvertTo-SecureString -String $clientSecret -AsPlainText -Force }
         if ($clientSecret -isnot [SecureString]) { throw "ClientSecret needs to be a SecureString or a String" }
 
-        $bcauthContext = New-BcAuthContext `
+        $script:bcauthContext = New-BcAuthContext `
             -clientID $clientID `
             -clientSecret $clientSecret `
             -tenantID $tenantId `
             -scopes "https://api.businesscentral.dynamics.com/.default"
 
-        if (-not ($bcAuthContext)) {
+        if (-not ($script:bcAuthContext)) {
             throw "Authentication failed"
         }
+    }
+    else {
+        $script:authContext = Renew-BcAuthContext -bcAuthContext $bcAuthContext
     }
 
     $appFolder = Join-Path ([System.IO.Path]::GetTempPath()) ([guid]::NewGuid().ToString())

--- a/Saas/Publish-PerTenantExtensionApps.ps1
+++ b/Saas/Publish-PerTenantExtensionApps.ps1
@@ -67,13 +67,13 @@ try {
         if ($clientSecret -is [String]) { $clientSecret = ConvertTo-SecureString -String $clientSecret -AsPlainText -Force }
         if ($clientSecret -isnot [SecureString]) { throw "ClientSecret needs to be a SecureString or a String" }
 
-        $script:bcauthContext = New-BcAuthContext `
+        $script:authContext = New-BcAuthContext `
             -clientID $clientID `
             -clientSecret $clientSecret `
             -tenantID $tenantId `
             -scopes "https://api.businesscentral.dynamics.com/.default"
 
-        if (-not ($script:bcAuthContext)) {
+        if (-not ($script:AuthContext)) {
             throw "Authentication failed"
         }
     }

